_posts/2018-11-13-type-constraints-correctness-reuse.md

```diff
@@ -0,0 +1,65 @@
+---
+layout: post
+title: Type constraints for code reuse and readability
+author: jguitana
+---
+
+The case for statically typed languages is so that we can prove our programs are right at compile time. Although our languages are not powerful enough to achieve this all the time, we certainly not always use their full capabilities. Our intuition should guide us to reduce implementation space and improve code reuse, allowing us to focus on small parts of a big problem. We can only do this if we choose the right abstraction in the form of Types, because these act as constraints on the implementation space, and if built correctly bring with them reusable and tested code.
+
+Let's take for example the `Int` and the `Nat` types. Int represents a number from `-x to x`. Nat represents a number from `0 to x`.
+
+*The crux of the question is*: Whenever we expect a natural number in some closure, is it different to pass an Int argument then it is passing a Nat argument?
+
+We are free to make a choice here, and many of us will not think about Nat.
+
+Or maybe just take Int for performance reasons? Well, not all abstractions are free of cost, but the same reason we give for not using lower level languages applies - we want to develop software faster.
+
+Maybe we will use negative numbers in the future? Well, first the program must run in the present. Refactoring constraints is ok, the compiler will help.
+
+And will our program be correct if we do use Int? It might, but we will have to test the case when Int goes negative. This is what is meant by reducing implementation space. Using Nat achieves this, but Int will not. So we end up by having to write extra tests. We can see that in a dynamic language with no types, everything is unconstrained to some primitive types, thus all implementations must be tested.
+
+We've made a quick mental exercise on the cases above and we can see downsides and upsides as with most things. The bottom line is that if we are not solving a C10k like problem, we are probably better off building a more readable, smaller, well constrained program.
+
+Let's now take a look at `List` and `NonEmptyList` (from cats library).
```

Comment:
I would add a hyperlink to cats documention on "cats" word.

Reply:
Agreed